solidity/contracts/bridge/Bridge.sol

```diff
@@ -264,34 +323,309 @@ contract Bridge {
+
+        // Inputs processing loop.
+        for (uint256 i = 0; i < inputsCount; i++) {
+            // Check if we are at the end of the input vector.
+            if (inputStartingIndex >= sweepTx.inputVector.length) {
+                break;
+            }
+
+            (bytes memory input, uint256 inputLength) =
+                extractTxInput(sweepTx, inputStartingIndex);
+
+            // Extract the transaction hash corresponding to the given input.
+            // Note that it's little-endian.
+            bytes32 inputTxHash = input.extractInputTxIdLE();
+
+            DepositInfo storage deposit =
+                unswept[
+                    uint256(
+                        keccak256(
+                            abi.encodePacked(
+                                inputTxHash,
+                                BTCUtils.reverseUint32(
+                                    uint32(input.extractTxIndexLE())
+                                )
+                            )
+                        )
+                    )
+                ];
+
+            if (deposit.revealedAt != 0) {
+                // TODO: Check if we can make the replay protection cheaper.
+                require(deposit.sweptAt == 0, "Deposit already swept");
+                /* solhint-disable-next-line not-rely-on-time */
+                deposit.sweptAt = uint32(block.timestamp);
+
+                depositors[processedDepositsCount] = deposit.depositor;
+                depositedAmounts[processedDepositsCount] = deposit.amount;
+                inputsValue += depositedAmounts[processedDepositsCount];
+
+                processedDepositsCount++;
+            } else if (
+                !previousSweepFlag && previousSweep.txHash == inputTxHash
+            ) {
+                inputsValue += previousSweep.amount;
+                previousSweepFlag = true;
+            } else {
+                revert("Unknown input type");
+            }
+
+            // Make the `inputStartingIndex` pointing to the next input by
+            // increasing it by current input's length.
+            inputStartingIndex += inputLength;
+        }
+
+        // Assert the previous sweep flag is true which means previous sweep
+        // output was used as current sweep input or previous sweep has not
+        // occurred at all.
+        require(
+            previousSweepFlag,
+            "Previous sweep output not present in sweep transaction inputs"
+        );
+
+        return (inputsValue, depositors, depositedAmounts);
+    }
+
+    // TODO: Documentation. Mention that `txInfo` data should be validated outside.
+    function extractTxInput(
+        BitcoinTx.Info calldata txInfo,
+        uint256 inputStartingIndex
+    ) internal pure returns (bytes memory input, uint256 inputLength) {
+        // First, determine the remaining vector using current input
+        // starting index.
+        bytes memory remainingVector =
```

Comment:
Every time we want to extract another input we copy the whole remaining part of the input vector.
It would be great if we could optimise it. Perhaps we could change the `determineInputLength` function, so that it can work on the whole input vector, but we could pass a parameter there that indicates the start of the input which length we want to know?

Reply:
I've added this capability to `bitcoin-spv`: https://github.com/keep-network/bitcoin-spv/pull/3/commits/45dfb963d3777d3bb9ec42c92e6827113895c889#diff-8e482a4e5e5d4d6a42a321153e706c491f81a1e9bf0881b4607ef4f916132370R294